### DISCUSSION TITLE
How can i start multiple instances via Powershell

### DISCUSSION THREAD

**Discussion by TimoTasty** [ImageMagick/ImageMagick#3850] [Help]
Hello there,

currently i have a PS-Script running through our fileserver (Get-Childitem -recurse) with approx. over 1,2 million images.
After checking some things and sorting out Images with a acceptable "StorageScore" ((Height x Width) / Filesize) i run the image against this line in PS:

. $Magick $File -interlace Plane -quality 70% $File

Since this line is called within a foreach-loop the script waits for the image to be compressed.

The script is running for 5 days now and no end is in sight.

So i moved the actual compressing task to a PS-Job with the Start-Job cmdlet. Which basically creates a new PS-Session (which i limit to 8 concurrent running jobs) in the Background an runs the code. Within this code everything else works (writing to logs, manipulating EXIF with exiftool, etc.)

BUT the Imagemagick processing with the line from above compresses only every 10th image or so.

I seems that magick.exe cannot be called within every of this jobs.

Anybody got a hint for me?

PS: We are running on a Windows Server 2016 with 256GB RAM and 32 Cores. I want this thing to work hard for us ;)

Thank you very much

Timo

**Comment by snibgo** [comment]
What errors do you get from ImageMagick? Do you even trap errors and report them?

**Reply by TimoTasty** [reply]
No, currently i am not reporting any errors of ImageMagick. How du you suggest i sould do this?

**Reply by TimoTasty** [reply]
Ok, now i got a log File. Hope this helps!
[Magick-0.log](https://github.com/ImageMagick/ImageMagick/files/6694293/Magick-0.log)

**Comment by snibgo** [comment]
The log file has multiple entries like:
```
missing output filename `-debug' at CLI arg 7
```
What was your command? I suspect it has an error.

**Reply by TimoTasty** [reply]
Yeah, that was my fault. When i finally changed the log.xml File and deleted the -debug switch in the magick-command the line didn't appear anymore. I can provide you to logfiles, processing the same image in the same Folder, one whule using the working "serial"script and one using the non-working "parallel"-script.

[Magick-0_Parallel.log](https://github.com/ImageMagick/ImageMagick/files/6694633/Magick-0_Parallel.log)
[Magick-0_Serial.log](https://github.com/ImageMagick/ImageMagick/files/6694635/Magick-0_Serial.log)

The only difference in calling ImageMagick is calling the Image-Name variable in different ways, because using a variable inside a Job from "outside" of itself makes this mandatory.

Serial > Working

 . $Magick $FileFN -interlace Plane -quality 70% -debug $FileFN 

Parallel > not Working

 . $Magick $using:FileFN -interlace Plane -quality 70% -debug $using:FileFN 

The Variable itself is working, because i use it to write the name in a logfile just 2 lines below...

**Comment by snibgo** [comment]
Okay, I suppose the problem was that your command ended with "-debug" instead of the output filename.

Comparing the serial and parallel logs: the parallel process runs "magick", but the does nothing else. The serial process runs "magick", which reads a JPG and does some work and writes a JPG.

I don't use Powershell, so can only guess at the problem. Your parallel command is:
```
. $Magick $using:FileFN -interlace Plane -quality 70% -debug $using:FileFN
```
What is a typical value of "$using:FileFN"?

**Reply by TimoTasty** [reply]
Ok, so i am one step further. I extracted the code from the start-job section and worked on it separately to get the console output.
From there a saw this line:
magick.exe : magick.exe: unable to open image 'C:\temp\1\23052007_ROBEND': No such file or directory @ error/blob.c/OpenBlob/3537.

The Filename contains a space. If i call the variable directly it is resolving to the full path. Strange behaviour. Especially because i use the exact same varriable in the working script.
But this is cleary a PS issue, as ImageMagick is working perfectly if i just eleminate the space!

I have to further investigate on my own. But thanks for your help!

Kind regards, Timo

**Comment by snibgo** [comment]
Try quoting the filenames, eg:
```
. $Magick "$using:FileFN" -interlace Plane -quality 70% -debug "$using:FileFN"
```

**Comment by TimoTasty** [comment]
I ultimately got it running!
All i had to do is write the $using Variable into a somehow "script-native" variable with $FileFullName = $_.Fullname

For everyone after us reading this i attached the whole script. Please be aware, that you definitely want to test this before running over a large scale storage like fileservers! I of course you use this at your own risk. 

Thanks for your help
[Image Compression_FINAL.zip](https://github.com/ImageMagick/ImageMagick/files/6701207/Image.Compression_FINAL.zip)

**Reply by TimoTasty** [reply]
Final Edit:

Add

Add-Type -AssemblyName System.Drawing

in Line 10. This way you can call the Script from Taskscheduler. Outside of the ISE the Script is much faster
